src/modules/mavlink/streams/BATTERY_STATUS.hpp

```diff
@@ -114,28 +114,39 @@ class MavlinkStreamBatteryStatus : public MavlinkStream
 					bat_msg.temperature = INT16_MAX;
 				}
 
-				static constexpr int mavlink_cells_max = (sizeof(bat_msg.voltages) / sizeof(bat_msg.voltages[0]));
-				static constexpr int uorb_cells_max =
-					(sizeof(battery_status.voltage_cell_v) / sizeof(battery_status.voltage_cell_v[0]));
+				// fill cell voltages
+				static constexpr uint8_t mavlink_cell_slots = (sizeof(bat_msg.voltages) / sizeof(bat_msg.voltages[0]));
+				static constexpr uint8_t mavlink_cell_slots_extension
+					= (sizeof(bat_msg.voltages_ext) / sizeof(bat_msg.voltages_ext[0]));
+				uint16_t cell_voltages[mavlink_cell_slots + mavlink_cell_slots_extension];
+
+				for (auto &voltage : cell_voltages) {
+					voltage = UINT16_MAX;
+				}
+
+				// We don't know the cell count or we don't know the indpendent cell voltages so we report the total voltage in the first cell
+				if (battery_status.cell_count == 0 || battery_status.voltage_cell_v[0] < 0.0001f) {
+					cell_voltages[0] = battery_status.voltage_filtered_v * 1000.f;
 
-				for (int cell = 0; cell < mavlink_cells_max; cell++) {
-					if (battery_status.connected && (cell < battery_status.cell_count) && (cell < uorb_cells_max)) {
-						bat_msg.voltages[cell] = battery_status.voltage_cell_v[cell] * 1000.f;
+				} else {
+					static constexpr uint8_t uorb_cell_slots =
+						(sizeof(battery_status.voltage_cell_v) / sizeof(battery_status.voltage_cell_v[0]));
 
-					} else {
-						bat_msg.voltages[cell] = UINT16_MAX;
+					for (uint8_t cell = 0; cell < mavlink_cell_slots + mavlink_cell_slots_extension; cell++) {
+						if (battery_status.connected && cell < math::min(battery_status.cell_count, uorb_cell_slots)) {
+							cell_voltages[cell] = battery_status.voltage_cell_v[cell] * 1000.f;
+						}
```

Comment:
Do you think if this can spare CPU cycles since both mavlink_cells and uorb_cells are predefined? Also **for loop** looks cleaner (maybe not with GitHub suggestion formatting xD)

```suggestion
					static constexpr uint8_t cell_slots =
							(mavlink_cell_slots + mavlink_cell_slots_extension) < uorb_cell_slots ?
							(mavlink_cell_slots + mavlink_cell_slots_extension) : uorb_cell_slots;

					for (uint8_t cell = 0; battery_status.connected && (cell < math::min(battery_status.cell_count, cell_slots)); cell++) {
						cell_voltages[cell] = battery_status.voltage_cell_v[cell] * 1000.f;
					}
```

Reply:
Good input, I took over the structure from before where every field had to be filled either with a value or with `UINT16_MAX` but the "cell not present" value is now set at initialization.

We only need to iterate the loop for number_of_cells = min(cell_count, uorb_slots, mavlink_slots).

I changed it in: aa38fdc